signac/core/synced_collection.py

```diff
@@ -106,28 +122,62 @@ def _load(self):
         pass
 
     @abstractmethod
-    def _sync(self):
+    def _sync(self, data):
         """Write data to underlying backend."""
         pass
 
+    @abstractmethod
+    def write_to_cache(self, data=None):
+        """Write data to cache."""
+        pass
+
+    @abstractmethod
+    def read_from_cache(self):
+        """Read data from cache."""
+        pass
+
+    def _sync_to_backend(self):
+        self._sync()
+
     def sync(self):
         """Synchronize the data with the underlying backend."""
         if self._suspend_sync_ <= 0:
             if self._parent is None:
-                self._sync()
+                self._sync_to_backend()
+                self.write_to_cache()
             else:
                 self._parent.sync()
 
     def load(self):
         """Load the data from the underlying backend."""
         if self._suspend_sync_ <= 0:
             if self._parent is None:
-                data = self._load()
+                data = self.read_from_cache()
+                if data is None:
+                    data = self._load()
+                    self.write_to_cache(data)
                 with self._suspend_sync():
                     self._update(data)
             else:
                 self._parent.load()
 
+    @contextmanager
+    def buffered(self):
+        """Context manager for buffering read and write operations.
+
+        This context manager activates the "buffered" mode, which
+        means that all read operations are cached, and all write operations
+        are deferred until the buffered mode is deactivated.
+
+        Yields
+        ------
+        buffered_collection : object
+            Buffered SyncedCollection object of corresponding base type.
+        """
+        buffered_collection = self.from_base(data=self, backend='buffered', parent=self)
+        yield buffered_collection
+        buffered_collection.flush()
```

Comment:
I vaguely recall something like this being in the old implementation? It may prevent a loss of data if an error occurs in the context manager. The goal is to guarantee the final flush occurs. I recommend looking over other parts of the caching/buffering implementation to look for similar patterns where this might be an issue.
```suggestion
        try:
            yield buffered_collection
        finally:
            buffered_collection.flush()
```